### PR TITLE
Add Custom Fields to Product Model in Networking

### LINKED ISSUE #13751: Update MetaData entity to include Product row
This needs https://github.com/woocommerce/woocommerce-ios/pull/13736 to be merged first

### PR DESCRIPTION
**Pull request by pmusolino**
Closes: #13751 part of #13493

## Summary description of changes

This PR introduces custom fields to the product model, improving the flexibility of metadata handling. The changes include updates to the product struct, encoding logic, and associated mock objects to accommodate these custom fields, which will be used later. Right now, the custom fields are not used in the code, and also the Storage/Yosemite layer implementation is missing and will be handled later on.

In future PRs, AddOns and subscription that currently use metadata, should be also refactored.

## Summary of changes
- Adds a `customFields` property to the `Product` model, allowing for additional metadata to be associated with products.
- Updates the encoding process to include the `customFields` when serializing the `Product` object.
- Modifies various files to ensure that `customFields` are initialized and handled correctly across the system.
- Introduces relevant TODO comments for future implementation of custom field handling within the storage layer.

## Testing
- I verified that products works like before.
- I did a smoke testing to understand if AddOns and Products with subscriptions works like before.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

File: Networking/Networking/Model/Product/Product.swift [hunk 1]
```diff
@@ -521,6 +525,9 @@ public struct Product: Codable, GeneratedCopiable, Equatable, GeneratedFakeable
 
         let menuOrder = try container.decode(Int.self, forKey: .menuOrder)
 
+        // Filter out metadata if the key is prefixed with an underscore (internal meta keys)
+        let customFields = (try? container.decode([MetaData].self, forKey: .metadata).filter({ !$0.key.hasPrefix("_")})) ?? []
```
Comment (selanthiraiyan) [inline Networking/Networking/Model/Product/Product.swift:529]: I think we should add unit tests to ensure that `customFields` are mapped as expected. Probably to `ProductMapperTests`. What do you think?

**Review by selanthiraiyan — COMMENTED** [review]
I smoke testes products tab in general, and it works as before. ✅ 

I think we should add unit tests to ensure that `customFields` are parsed as expected from a mock JSON. Please correct me if I am missing any context here.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/222">20.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Networking/Networking/Model/Product/Product.swift (hunk 1 above)
Comment (pmusolino) [reply Networking/Networking/Model/Product/Product.swift:529]: Good point. I implemented this test: ff9b457560f3ea7ff728b7b8ce0ac01115da2cd6

**Review by selanthiraiyan — APPROVED** [review]
Thanks for adding the unit test, Paolo! The code looks good to me. ✅ 

The PR is ready to merge once we fix that failing unit test. 🚢

**Pull request merged**